packages/react/tooltip/src/Tooltip.tsx

```diff
@@ -330,16 +365,132 @@ interface TooltipContentProps extends TooltipContentImplProps {
+  const providerContext = useTooltipProviderContext(TOOLTIP_NAME, props.__scopeTooltip);
+  const ref = React.useRef<TooltipContentPointerAccessibleElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, ref);
+  const [pointerGraceArea, setPointerGraceArea] = React.useState<Polygon | null>(null);
+  const pointerGraceTimerRef = React.useRef(0);
+
+  const { onClose } = context;
+
+  const handleRemoveGraceArea = React.useCallback(() => {
+    setPointerGraceArea(null);
+    removeDebugArea();
+    window.clearTimeout(pointerGraceTimerRef.current);
+    providerContext.isPointerInTransitRef.current = false;
+  }, [providerContext.isPointerInTransitRef]);
+
+  const handleCreateGraceArea = React.useCallback(() => {
+    const content = ref.current;
+    const trigger = context.trigger;
+
+    if (content && trigger) {
+      const side = content.dataset.side as Side;
+      if (side) {
+        const triggerRect = trigger.getBoundingClientRect();
+        const contentRect = content.getBoundingClientRect();
+        const triggerEdges = getEdgesFromRect(triggerRect);
+        const contentEdges = getEdgesFromRect(contentRect);
+        const isXAxis = side === 'left' || side === 'right';
+        const [contentFarEdge, contentNearEdge] = contentEdges[side];
+        const [triggerFarEdge] = triggerEdges[side];
+
+        const shouldTrimNearEdge = isXAxis
+          ? contentRect.height < triggerRect.height
+          : contentRect.width < triggerRect.width;
+
+        const [startPoint, ...lastPoints] = shouldTrimNearEdge
+          ? contentFarEdge
+          : contentNearEdge.concat(contentFarEdge.reverse());
+
+        const graceArea = [startPoint, ...triggerFarEdge, ...lastPoints];
+        setPointerGraceArea(graceArea);
+        createDebugArea(graceArea);
+      }
+
+      pointerGraceTimerRef.current = window.setTimeout(() => {
+        onClose();
+        handleRemoveGraceArea();
+      }, 500);
+
+      providerContext.isPointerInTransitRef.current = true;
+    }
+  }, [context.trigger, onClose, handleRemoveGraceArea, providerContext.isPointerInTransitRef]);
+
+  React.useEffect(() => {
+    return () => handleRemoveGraceArea();
+  }, [handleRemoveGraceArea]);
+
+  React.useEffect(() => {
+    const trigger = context.trigger;
+    const content = ref.current;
+
+    if (trigger && content) {
+      trigger.addEventListener('pointermove', handleRemoveGraceArea);
+      content.addEventListener('pointermove', handleRemoveGraceArea);
+      trigger.addEventListener('pointerleave', handleCreateGraceArea);
+      content.addEventListener('pointerleave', handleCreateGraceArea);
+
+      return () => {
+        trigger.removeEventListener('pointermove', handleRemoveGraceArea);
+        content.removeEventListener('pointermove', handleRemoveGraceArea);
+        trigger.removeEventListener('pointerleave', handleCreateGraceArea);
+        content.removeEventListener('pointerleave', handleCreateGraceArea);
+      };
+    }
```

Comment:
LMK what you think about all of this, I wanted to co-locate as much of the behaviour as I could in single removable place.